common/djangoapps/student/admin.py

```diff
@@ -149,13 +150,26 @@ class Meta(object):
 class CourseEnrollmentForm(forms.ModelForm):
 
     def __init__(self, *args, **kwargs):
+        # If args is a QueryDict, then the ModelForm addition request came in as a POST with a course ID string.
+        # Change the course ID string to a CourseLocator object by copying the QueryDict to make it mutable.
+        if len(args) > 0 and 'course' in args[0] and isinstance(args[0], QueryDict):
+            args_copy = args[0].copy()
+            try:
+                args_copy['course'] = CourseKey.from_string(args_copy['course'])
+            except InvalidKeyError:
+                raise forms.ValidationError("Cannot make a valid CourseKey from id {}!".format(args_copy['course']))
+            args = [args_copy]
```

Comment:
Are we confident that there is no other data in args that should be preserved?

Reply:
I checked the parameters in my devstack. `args` is a tuple. On GET, `args` is empty so there's not even `args[0]`. On POST there's `args[0]`, and also `args[1]` which would be the attached files to this form, and since there aren't any in this form, it's always an empty dict that can be discarded. So as long as no FileFields are added to this form, `args` is preserved, and `kwargs` too.

The same happens with [this other usage of the code](https://github.com/edx/edx-platform/blob/4d37779f58f1132176fa5aa095692f01a479f769/common/djangoapps/course_modes/admin.py#L68), it will work as long as the form has no files.
We can merge it if we predict there won't be files, or we can add an explicit comment to those two usages, or we could make it handle the potential case with attached files. @e0d which one is best?

Details:

On GET (when opening the form, or when requesting the delete the object):
```
args:
()
kwargs:
{'instance': <CourseEnrollment: [CourseEnrollment] audit: course-v1:edX+DemoX+Demo_Course (2017-06-07 00:44:32.887985+00:00); active: (False)>}
```

On POST (when changing the form):
```
args:
(<QueryDict: {u'_save': [u'Save'], u'is_active': [u'on'], u'course': [u'course-v1:edX+DemoX+Demo_Course'], u'user': [u'7'], u'csrfmiddlewaretoken': [u'zWJC2Jw2YvVubdwR9BRjVSgdyaXeyBwQztWSbGzSAQdrW7XRkwxLAy3vgWRS3Iey'], u'mode': [u'audit']}>, <MultiValueDict: {}>)
kwargs:
{'instance': <CourseEnrollment: [CourseEnrollment] audit: course-v1:edX+DemoX+Demo_Course (2017-06-07 00:44:32.887985+00:00); active: (False)>}
```

On POST (when deleting the object) it doesn't go through this code.

For context, this is how Django is initializing the form:

on GET:

```
…
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/utils/decorators.py", line 63, in bound_func
    return func.__get__(self, type(self))(*args2, **kwargs2)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/contrib/admin/options.py", line 1409, in changeform_view
    return self._changeform_view(request, object_id, form_url, extra_context)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/contrib/admin/options.py", line 1466, in _changeform_view
    form = ModelForm(instance=obj)
```

on POST:
```
…
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/utils/decorators.py", line 63, in bound_func
    return func.__get__(self, type(self))(*args2, **kwargs2)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/contrib/admin/options.py", line 1409, in changeform_view
    return self._changeform_view(request, object_id, form_url, extra_context)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/contrib/admin/options.py", line 1440, in _changeform_view
    form = ModelForm(request.POST, request.FILES, instance=obj)
```